rpc/src/client/subscription.rs

```diff
@@ -0,0 +1,591 @@
+
+use crate::client::sync::{unbounded, ChannelRx, ChannelTx};
+use crate::client::ClosableClient;
+use crate::event::Event;
+use crate::{Error, Id, Result};
+use async_trait::async_trait;
+use futures::task::{Context, Poll};
+use futures::Stream;
+use getrandom::getrandom;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::pin::Pin;
+
+/// A client that exclusively provides [`Event`] subscription capabilities,
+/// without any other RPC method support.
+///
+/// [`Event`]: event/struct.Event.html
+#[async_trait]
+pub trait SubscriptionClient: ClosableClient {
+    /// `/subscribe`: subscribe to receive events produced by the given query.
+    ///
+    /// For query syntax details, see the
+    /// [Tendermint RPC docs](https://docs.tendermint.com/master/rpc/#/Websocket/subscribe).
+    async fn subscribe(&mut self, query: String) -> Result<Subscription>;
+}
+
+/// An interface that can be used to asynchronously receive [`Event`]s for a
+/// particular subscription.
+///
+/// ## Examples
+///
+/// ```
+/// use tendermint_rpc::{SubscriptionId, Subscription};
+/// use futures::StreamExt;
+///
+/// /// Prints `count` events from the given subscription.
+/// async fn print_events(subs: &mut Subscription, count: usize) {
+///     let mut counter = 0_usize;
+///     while let Some(res) = subs.next().await {
+///         // Technically, a subscription produces `Result<Event, Error>`
+///         // instances. Errors can be produced by the remote endpoint at any
+///         // time and need to be handled here.
+///         let ev = res.unwrap();
+///         println!("Got incoming event: {:?}", ev);
+///         counter += 1;
+///         if counter >= count {
+///             break
+///         }
+///     }
+/// }
+/// ```
+///
+/// [`Event`]: ./event/struct.Event.html
+#[derive(Debug)]
+pub struct Subscription {
+    /// The query for which events will be produced.
+    pub query: String,
+    /// The ID of this subscription (automatically assigned).
+    pub id: SubscriptionId,
+    // Our internal result event receiver for this subscription.
+    event_rx: ChannelRx<Result<Event>>,
+    // Allows us to gracefully terminate this subscription.
+    terminate_tx: ChannelTx<TerminateSubscription>,
+}
+
+impl Stream for Subscription {
+    type Item = Result<Event>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.event_rx.poll_recv(cx)
+    }
+}
+
+impl Subscription {
+    pub(crate) fn new(
+        id: SubscriptionId,
+        query: String,
+        event_rx: ChannelRx<Result<Event>>,
+        terminate_tx: ChannelTx<TerminateSubscription>,
+    ) -> Self {
+        Self {
+            id,
+            query,
+            event_rx,
+            terminate_tx,
+        }
+    }
+
+    /// Gracefully terminate this subscription.
+    ///
+    /// This can be called from any asynchronous context. It only returns once
+    /// it receives confirmation of termination.
```

Comment:
It consumes the `Subscription`, so there's no way it can be called from two contexts simultaneously. I'll reword this to make it clearer - what I actually meant was that the `Subscription` can be moved to any async context and it can be terminated within that context.